### PR TITLE
[WIP] Portable install

### PR DESCRIPTION
**Pull request by sblaisot**
This PR add a portable mode to Mixxx (lp:1731689).

The portable mode is triggered if Mixxx executable is named `mixxx-portable` (case-insensitive and without extension, thus working with `Mixxx-portable` under MacOSX and `mixxx-portable.exe` under windows).
The portable mode can also be triggered with new command-line argument `--portable`

When in portable mode, Mixxx will not try to access (for reading or writing) to resources or settings outside of specified directories.
Default settings directory in portable mode is a subdirectory `mixxx-settings` in the same place where the Mixxx-portable executable resides. This can be overwritten with the usual `--settingsPath` command-line argument.
Default resources directory in portable mode is a subdirectory `mixxx-resources` in the same place where the Mixxx-portable executable resides. This can be overwritten with the usual `--resourcePath` command-line argument.

In portable mode, no attempt will be made to migrate old settings from standard folders to avoid any interaction outside Mixxx directory

I also added a `makeportable` target to scons producing a portable archive using 7-zip and configured appveyor to produce portable archives and store them as artifact.

This PR doesn't address Music/tracks folders. This will be added in a future PR

### CODE REVIEW

**Comment by sblaisot** [comment]
I should probably add CHANGELOG, README and LICENCE in the archive. Will add that tomorrow.

**Comment by daschuer** [comment]
Nice Idea. I hope I will find time to test this soon.

 Is such magic a standard behaviour among other windows Apps?

Is there anything else to do? I think the use case is to install Mixxx an a jump drive along with all the tracks. Does it work flavlessly when the drive letter changes?

**Comment by sblaisot** [comment]
AFAIK there is no standard behaviour about portable apps. often, this is done by a launchscript which prepare the configuration before launching and cleanup after closing the application.

I did this that way for settingsPath because we already had that logic for RessourcePath (with another use-case but similar intent).

We currently don't handle relative tracks so this will be a problem when changing drive letter. Il will try ti address this later in another PR.

**Comment by sblaisot** [comment]
> I think the use case is to install Mixxx an a jump drive along with all the tracks. Does it work flavlessly when the drive letter changes?

Maybe we can handle that by replacing the `"[Playlist]", "Directory"` configkey whet Mixxx is detects a portable execution to point to a "tracks" subdir near the Mixxx executable if it exists.
But doing this will prevent a user to be prompted at first launch to select a Music directory.
We should then explain to portable users how to add another library folder if he wants.
Seems doable, though.

**Comment by sblaisot** [comment]
> Neat, thanks for working on this. I am wondering if maybe we should use a more descriptive magic directory name than "settings" like maybe "mixxx-settings".

Yes, this is possible ;)

> I could imagine a user putting unrelated files and directories in the same folder as the Mixxx executable and accidentally deleting or renaming a directory simply called "settings" because they do not know what it is.

I see a far bigger risk with the `res` sudbirectory (but this one can also be changed to `mixxx-res` if we want, but also with the `plugins` dir and we have no control on the name of that directory I'm afraid.

**Comment by Pegasus-RPG** [comment]
Not trying to bike-shed here, but maybe it's better to have a dir called 'portable' then 'settings' and 'resources' in it to keep things neater and more obvious? (That also future-proofs it.)

**Comment by sblaisot** [comment]
> maybe it's better to have a dir called 'portable' then 'settings' and 'resources' in it

I was thinking about that also. I will try that also.

**Comment by daschuer** [comment]
> I am wondering if maybe we should use a  more descriptive magic directory name than "settings" like maybe  "mixxx-settings". I could imagine a user putting unrelated files and  directories in the same folder as the Mixxx executable and accidentally  deleting or renaming a directory simply called "settings" because they  do not know what it is.

If we want to protect against this, all additional folders needs the mixxx prefix. If we cannot effort this, I would vote for the original folder names without prefix.

Sometimes it happens to me that I extract a zip achive accidentally to my download folder without a subfolder. This way, I have the exe, various dlls and ini files among other downloaded files, makeing it hard to undo the mistake . 
I think we can save us for this by using a mixxx folder and just adding a single short cut outside this folder. 

@Be-ing does this also help in your use case?

**Review by Pegasus-RPG — CHANGES_REQUESTED** [review]

File: src/SConscript [hunk 1]
```diff
@@ -1195,3 +1195,104 @@ env.Append(BUILDERS = {'BuildUbuntuPackage' : versiondebbld})
...
+    print('  . Copying documentation:')
+    for file in ['CHANGELOG', 'LICENCE', 'build/portable/README.md']:
+        print('    - {}'.format(os.path.basename(str(file))))
+        shutil.copy2('{}/{}'.format(dist_dir, os.path.basename(str(file))),
+                     'portable/{}/{}'.format(directory_name, os.path.basename(str(file))))
+
+    print("Now building portable package...")
+    # Zip the directory_name into target archive file
+    os.chdir('portable')
+    os.system('7z a ../{} {}'.format(archive_name, directory_name))
```
[the remark this answers is not in the document]
Comment (sblaisot) [reply src/SConscript]: 7-zip is smart enough. But if we want to make sure, I can add `-tzip`

```
-t{Type}: Set type of archive
```

Not sure it is a good idea to make an installer for a portable application ;)

[the remark this answers is not in the document]
Comment (Pegasus-RPG) [reply src/SConscript]: Oh, no, not an installer. An SFX just runs a little application asking where they want to extract the files to.

[the remark this answers is not in the document]
Comment (Pegasus-RPG) [reply src/SConscript]: And if it's smart enough, then just leave it, incase we decide to change the format in the future, we only have to do it in one place.

[the remark this answers is not in the document]
Comment (sblaisot) [reply src/SConscript]: It is possible to make a self-extracting archive with a nice GUI under windows with

```
7z a mixxx.exe -sfx7z.sfx mixxx-2.2.0-alpha-pre-portable_install-git6548-release-x64
```

But I need to test it under linux and MacOSX to verify it is cross-platform

File: src/preferences/upgrade.cpp [hunk 2]
```diff
@@ -47,127 +47,130 @@ Upgrade::~Upgrade() {
...
-        }
+            if (!newLocation.exists()) {
+                qDebug() << "Creating new settings directory" << newLocation.absolutePath();
+                newLocation.mkpath(".");
+            }
 
-        QString errorText = "Error moving your %1 file %2 to the new location %3: \n";
+            QString errorText = "Error moving your %1 file %2 to the new location %3: \n";
 
 #ifdef __WINDOWS__
```
[the remark this answers is not in the document]
Comment (sblaisot) [reply src/preferences/upgrade.cpp]: done

File: src/preferences/upgrade.cpp [hunk 3]
```diff
@@ -47,127 +47,130 @@ Upgrade::~Upgrade() {
...
+    *
+    *   Since we didn't store version numbers in the config file prior to 1.7.0,
+    *   we check to see if the user is upgrading if his config files are in the old location,
+    *   since we moved them in 1.7.0. This code takes care of moving them.
+    */
+
+    // If we are running portable, do not try reading configuration outside our directory
+    if (! CmdlineArgs::Instance().getIsPortable()) {
+        QDir oldLocation = QDir(QDir::homePath());
 #ifdef __WINDOWS__
```
[the remark this answers is not in the document]
Comment (sblaisot) [reply src/preferences/upgrade.cpp]: done

File: src/SConscript (hunk 1 above)
[the remark this answers is not in the document]
Comment (sblaisot) [reply src/SConscript]: just putting this here for future reference : https://netbeansscribbles.wordpress.com/2015/01/30/creating-a-self-extracting-7zip-archive-multi-platform/

**Comment by sblaisot** [comment]
ok, first portable version available for test here: https://ci.appveyor.com/project/mixxxdj/mixxx/build/master-4951/job/0u9msvalom50m489/artifacts
(windows, 64 bits)

**Comment by Be-ing** [comment]
>  I think we can save us for this by using a mixxx folder and just adding a single short cut outside this folder.

Yes, I like this idea. Perhaps instead of checking for special directory names in the same directory as the executable, we can check for the executable being in a special directory called "mixxx" or "mixxx-portable". If the executable is in that special directory, look for "settings" and "res" in that directory too.

> I think the use case is to install Mixxx an a jump drive along with all the tracks. Does it work flavlessly when the drive letter changes?

I think we need to solve this for this PR to be really useful. It might be better to do that in another PR.

It would be awesome if we could get this to work with an AppImage on Linux and if it works on macOS. Ideally a user could put their music files on a USB drive and have a "mixxx-portable" folder with the Windows exe, macOS .app, and Linux AppImage as a backup in case their laptop fails at a gig. This would be a lot easier to maintain than a live distro just for Mixxx.

**Comment by sblaisot** [comment]
> we can check for the executable being in a special directory called "mixxx" or "mixxx-portable".

Going that way, I prefer to check wether the executable itself is called mixxx-portable or not.

>> I think the use case is to install Mixxx an a jump drive along with all the tracks. Does it work flavlessly when the drive letter changes?

> I think we need to solve this for this PR to be really useful. It might be better to do that in another PR.

I agreee this should be solved, but I prefer to work on it in another PR. This is not trivial at first look.

> It would be awesome if we could get this to work with an AppImage on Linux and if it works on macOS. 

Did you notice this PR is OS-agnostic and produces a portable archive for linux and macOSX as well ?
Just `scons makeportable`. 

That said, portable archive are less usable on linux because
1. we don't build statically so the system running Mixxx should have all dependencies installed
2. even if we build statically, we always depend on glibc version which can be different between distributions.
I don't know about macOSX on this subject.

**Comment by Be-ing** [comment]
> Going that way, I prefer to check wether the executable itself is called mixxx-portable or not.

I think that's a better idea. Changing program behavior with the name of the executable is occasionally used by other programs for example BusyBox. On the other hand I do not know of any program that relies on special directory names.

> Did you notice this PR is OS-agnostic and produces a portable archive for linux and macOSX as well ?

I did not notice, thank you for clarifying.

> That said, portable archive are less usable on linux because
>    we don't build statically on linux so the system running Mixxx should have all dependencies installed
>    even if we build statically, we always depend on glibc version which can be different between distributions.

Those can both be resolved by packaging an AppImage built on an old distro.

File: src/SConscript [hunk 4]
```diff
@@ -1195,3 +1195,104 @@ env.Append(BUILDERS = {'BuildUbuntuPackage' : versiondebbld})
...
+        shutil.copytree(src, dest)
+    # Directories are the same
+    except shutil.Error as e:
+        print('Directory not copied. Error: %s' % e)
+    # Any error saying that the directory doesn't exist
+    except OSError as e:
+        print('Directory not copied. Error: %s' % e)
+
+def BuildPortablePackage(target, source, env):
+    print("==== Mixxx Post-Build Checks ====")
```
[the remark this answers is not in the document]
Comment (sblaisot) [reply src/SConscript]: Done

**Comment by sblaisot** [comment]
> Those can both be resolved by packaging an AppImage built on an old distro.

Yes but that's beyond the scope of this PR ;)

I will take a look on portable behaviour based on executable name

**Comment by sblaisot** [comment]
ok, with last commit, Mixxx switch to portable mode if the executable itself is called `mixxx-portable` (case insensitive and without extension) or if the `--portable` switch is used on command-line.

When in portable mode, it uses by default `mixxx-settings` and `mixxx-resources` as settings an resources directories respectively. These can be overwriten with standard `--settingsPath` and `--resourcePath`

PR description updated

**Comment by sblaisot** [comment]
Argh, it doesn't work because QApplication is not even instanciated when parsing command line args so I can't use `QCoreApplication::applicationFilePath` to get Mixxx executable name :(

All I get is `QCoreApplication::applicationFilePath: Please instantiate the QApplication object first`

I need to find another way to get Mixxx executable name...

**Comment by sblaisot** [comment]
Leaving this here for reference:
* http://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qcoreapplication.cpp#n2227
* https://github.com/mixxxdj/mixxx/blob/master/src/main.cpp#L85-L100

**Comment by radusuciu** [comment]
Could you instantiate a dummy QCoreApplication just to get access to that method? Or access directly from the args variable?

**Comment by Symbian9** [comment]
> > Those can both be resolved by packaging an AppImage built on an old distro.
>
> Yes but that's beyond the scope of this PR ;)

But has you plans to make *AppImage* builds? I really want it use portable *Mixxx* on Linux.

**Pull request closed without merging**